spec/models/product_spec.rb

```diff
@@ -0,0 +1,142 @@
+require "rails_helper"
+
+RSpec.describe Product, type: :model do
+  subject {product}
+
+  let!(:supplier) { FactoryBot.create :supplier }
+  let!(:category) { FactoryBot.create :category }
+  let!(:product) {
+    FactoryBot.build :product,
+      supplier_id: supplier.id, category_id: category.id
+  }
+
+  describe "associations" do
+    it "belong to category" do
+      is_expected.to belong_to :category
+    end
+
+    it "belong to supplier" do
+      is_expected.to belong_to :supplier
+    end
+
+    it "has many image" do
+      is_expected.to have_many :images
+    end
+
+    it "has many order_details" do
+      is_expected.to have_many :order_details
+    end
+
+    it "has many comments" do
+      is_expected.to have_many :comments
+    end
+  end
+
+  describe "Delegate" do
+    it { is_expected.to delegate_method(:name).to(:category).with_prefix(true) }
+  end
+
+  describe "Validations" do
+    before {product.save}
+
+    it "valid all field" do
+      expect(subject).to be_valid
+    end
+
+    [:name, :description,:inventory_number, :price].each do |field|
+      it {
+        is_expected.to validate_presence_of(field)
+      }
+    end
+
+    it {
+      is_expected.to validate_length_of(:name)
+        .is_at_most(Settings.validate.products.max_length_name)
+    }
+
+    it {
+      is_expected.to validate_length_of(:description)
+        .is_at_most(Settings.validate.products.max_length_description)
+    }
+
+    it {
+      is_expected.to validate_numericality_of(:inventory_number)
+    }
+
+    it {
+      is_expected.to validate_numericality_of(:price)
+    }
+  end
+
+  describe "Scopes" do
+    let!(:r1) {
+      FactoryBot.create :product, name: "may khoan",
+        price: 100000, supplier_id: supplier.id, category_id: category.id
+    }
+    let!(:r2) {
+      FactoryBot.create :product, name: "may khoan2",
+        price: 170000, supplier_id: supplier.id, category_id: category.id
+    }
+    let!(:r3) {
+      FactoryBot.create :product, name: "bua ta",
+        price: 120000, supplier_id: supplier.id, category_id: category.id
+    }
+    let!(:r4) {
+      FactoryBot.create :product, name: "bua ta nho",
+        price: 190000, supplier_id: supplier.id, category_id: category.id
+   }
+
+    context "with supplier filtered" do
+      it "should return filtered products by supplier" do
+        expect(Product.by_supplier(supplier.id).pluck(:supplier_id))
+          .to eq [r1.supplier_id, r2.supplier_id,
+             r3.supplier_id, r4.supplier_id]
```

Comment:
sao đoạn này cồng kềnh thế e 
Product.by_supplier(supplier.id).to eq [r1, r2,r3, r4]

Reply:
đoạn này em cố tình cho nó ra tất cả những thằng có supplier_id thì viết nó dài

Reply:
không cần viết dài ra thế đâu e

Reply:
đọc dài khó hiểu lắm e